### PR TITLE
Support multiple email-sending domains

### LINKED ISSUE #127: Support multiple email-sending domains per assessment
# 💡 Summary #

Enable an assessment to have more than one email-sending domain.

## Motivation and context ##

Some types of assessments require multiple instances (e.g. Teamservers), where each instance sends email from a different domain.  Currently, our `email_sending_domain` variable only supports a single domain.

## Implementation notes ##

Convert the `email_sending_domain` variable from a string to a list.

## Acceptance criteria ##

- [ ] Terraform and cloud-init function correctly with no email-sending domains, a single Teamserver, and a single Gophish instance
- [ ] Terraform and cloud-init function correctly with a single email-sending domain, a single Teamserver, and a single Gophish instance
- [ ] Terraform and cloud-init function correctly with two email-sending domains, two Teamservers, and two Gophish instances
- [ ] Documentation explains how list of email-sending domains is applied to instances that send mail (teamserver, gophish).

### PR DESCRIPTION
**Pull request by dav3r**
#  #

## 🗣 Description ##

This PR enables support for multiple email-sending domains by Gophish and Teamserver instances in an assessment environment.

## 💭 Motivation and context ##

Assessment teams require the ability to have multiple Teamservers deployed in a single assessment environment, each configured with their own email-sending domain.  This is useful if one domain can no longer be feasibly used (e.g. it gets blocked), then the team can quickly switch over to using a different domain.

Both Teamservers and Gophish instances follow the same logic for utilizing domains:

Instances are deployed with each sequential domain in the provided list of email-sending domains, so `teamserver0` and `gophish0` will get the first domain, `teamserver1` and `gophish1` will get the second domain, and so on.  If there are more Teamserver or Gophish instances than email-sending domains, the domains in the list will be reused in a wrap-around fashion. For example, if there are three Teamservers deployed and only two email-sending domains, `teamserver0` will get the first domain, `teamserver1` will get the second domain, and `teamserver2` will wrap-around back to using the first domain.

Resolves #127 and resolves cisagov/cool-system-internal#42.

## 🧪 Testing ##

I tested these changes by deploying a varying number of Gophish and Teamserver instances, along with a varying number of email-sending domains, then confirming that the expected domain was applied to each instance.

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

File: gophish_cloud_init.tf [hunk 1]
```diff
@@ -96,15 +96,21 @@ data "cloudinit_config" "gophish_cloud_init_tasks" {
...
-        cert_bucket_name    = var.cert_bucket_name
-        cert_read_role_arn  = module.email_sending_domain_certreadrole.role.arn
+        aws_region       = var.aws_region
+        cert_bucket_name = var.cert_bucket_name
+        # We use the element() function below instead of the built-in list
+        # index syntax because we want the "wrap-around" behavior provided
+        # by element().  This means that the number of items in
+        # var.email_sending_domains does not have to exactly match the number
+        # of Gophish instances.  For details, see:
+        # https://www.terraform.io/docs/language/functions/element.html
```
Comment (jsf9k) [inline gophish_cloud_init.tf:106]: Nice comment!

File: README.md [hunk 2]
```diff
@@ -505,8 +505,8 @@ the COOL environment.
 | debian\_desktop\_security\_group | The security group for the Debian desktop instances. |
 | efs\_client\_security\_group | A security group that should be applied to all instances that will mount the EFS file share. |
 | efs\_mount\_targets | The mount targets for the EFS file share. |
-| email\_sending\_domain\_certreadrole | The IAM role that allows for reading the certificate for the email-sending domain. |
-| gophish\_instance\_profile | The instance profile for the Gophish instances. |
+| email\_sending\_domain\_certreadrole | The IAM roles that allow for reading the certificate for the email-sending domain. |
+| gophish\_instance\_profile | The instance profiles for the Gophish instances. |
```
Comment (jsf9k) [inline README.md]: Why do these outputs have singular names?  I think plural names (e.g. `email_sending_domain_certreadroles`) would make more sense, and would be more consistent with the input variable `email_sending_domains` above.

File: README.md [hunk 3]
```diff
@@ -534,7 +534,7 @@ the COOL environment.
 | samba\_server\_security\_group | The security group for the Samba file share server instances. |
 | scanner\_security\_group | A security group that should be applied to all instance types that perform scanning.  This security group allows egress to anywhere as well as ingress from anywhere via ICMP. |
 | ssm\_session\_role | An IAM role that allows creation of SSM SessionManager sessions to any EC2 instance in this account. |
-| teamserver\_instance\_profile | The instance profile for the Teamserver instances. |
+| teamserver\_instance\_profile | The instance profiles for the Teamserver instances. |
```
Comment (jsf9k) [inline README.md]: See previous comment in this same file.

**Review by jsf9k — COMMENTED** [review]
Please consider my comments regarding giving a few output variables names that reflect the plurality of the values they contain.

File: README.md (hunk 2 above)
Comment (dav3r) [reply README.md]: Agreed - please see https://github.com/cisagov/cool-assessment-terraform/pull/153/commits/80a13efd98e839e727fc46b81bc8a35fb4d421fc.

File: README.md (hunk 3 above)
Comment (dav3r) [reply README.md]: 👍   See https://github.com/cisagov/cool-assessment-terraform/pull/153/commits/80a13efd98e839e727fc46b81bc8a35fb4d421fc.

**Review by jsf9k — APPROVED** [review]
Thanks for taking my comments into consideration @dav3r.  Strong work here.

File: gophish_iam.tf [hunk 4]
```diff
@@ -3,50 +3,62 @@
 
 # The instance profile to be used
```
Comment (mcdonnnj) [inline gophish_iam.tf]:
```suggestion
# The instance profiles to be used
```

File: gophish_iam.tf [hunk 5]
```diff
@@ -3,50 +3,62 @@
...
+
   provider = aws.provisionassessment
 
-  name = "gophish_instance_profile_${terraform.workspace}"
-  role = aws_iam_role.gophish_instance_role.name
+  name = "gophish${count.index}_instance_profile_${terraform.workspace}"
+  role = aws_iam_role.gophish_instance_role[count.index].name
 }
 
 # The instance role
```
Comment (mcdonnnj) [inline gophish_iam.tf]:
```suggestion
# The instance roles
```

File: gophish_iam.tf [hunk 6]
```diff
@@ -3,50 +3,62 @@
...
   provider = aws.provisionassessment
 
   name   = "assume_delegated_role_policy"
-  role   = aws_iam_role.gophish_instance_role.id
-  policy = data.aws_iam_policy_document.gophish_assume_delegated_role_policy_doc.json
+  role   = aws_iam_role.gophish_instance_role[count.index].id
+  policy = data.aws_iam_policy_document.gophish_assume_delegated_role_policy_doc[count.index].json
 }
 
 # Attach the CloudWatch Agent policy to this role as well
```
Comment (mcdonnnj) [inline gophish_iam.tf]:
```suggestion
# Attach the CloudWatch Agent policy to these roles as well
```

File: README.md [hunk 7]
```diff
@@ -505,8 +505,8 @@ the COOL environment.
 | debian\_desktop\_security\_group | The security group for the Debian desktop instances. |
 | efs\_client\_security\_group | A security group that should be applied to all instances that will mount the EFS file share. |
 | efs\_mount\_targets | The mount targets for the EFS file share. |
-| email\_sending\_domain\_certreadrole | The IAM role that allows for reading the certificate for the email-sending domain. |
-| gophish\_instance\_profile | The instance profile for the Gophish instances. |
+| email\_sending\_domain\_certreadroles | The IAM roles that allow for reading the certificate for the email-sending domain. |
```
Comment (mcdonnnj) [inline README.md]: This one's just a thought.

```suggestion
| email\_sending\_domain\_certreadroles | The IAM roles that allow for reading the certificate for each email-sending domain. |
```

File: email_sending_domain_certreadrole.tf [hunk 8]
```diff
@@ -1,5 +1,7 @@
 # Create a role that allows an instance to read its certs from S3.
```
Comment (mcdonnnj) [inline email_sending_domain_certreadrole.tf]: Just mirroring the language used elsewhere in this repo/PR.

```suggestion
# Create roles that allow each instance to read its email-sending domain
# certificate from an S3 bucket.
```

File: gophish_iam.tf [hunk 9]
```diff
@@ -3,50 +3,62 @@
...
+  count = lookup(var.operations_instance_counts, "gophish", 0)
+
   provider = aws.provisionassessment
 
-  role       = aws_iam_role.gophish_instance_role.id
+  role       = aws_iam_role.gophish_instance_role[count.index].id
   policy_arn = "arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy"
 }
 
 # Attach the SSM Agent policy to this role as well
```
Comment (mcdonnnj) [inline gophish_iam.tf]:
```suggestion
# Attach the SSM Agent policy to these roles as well
```

File: outputs.tf [hunk 10]
```diff
@@ -58,14 +58,14 @@ output "efs_mount_targets" {
   description = "The mount targets for the EFS file share."
 }
 
-output "email_sending_domain_certreadrole" {
+output "email_sending_domain_certreadroles" {
   value       = module.email_sending_domain_certreadrole
-  description = "The IAM role that allows for reading the certificate for the email-sending domain."
+  description = "The IAM roles that allow for reading the certificate for the email-sending domain."
```
Comment (mcdonnnj) [inline outputs.tf]: Just mirrors the README suggestion for convenience.

```suggestion
  description = "The IAM roles that allow for reading the certificate for each email-sending domain."
```

File: teamserver_iam.tf [hunk 11]
```diff
@@ -3,50 +3,62 @@
 
 # The instance profile to be used
```
Comment (mcdonnnj) [inline teamserver_iam.tf]:
```suggestion
# The instance profiles to be used
```

File: teamserver_iam.tf [hunk 12]
```diff
@@ -3,50 +3,62 @@
...
+
   provider = aws.provisionassessment
 
-  name = "teamserver_instance_profile_${terraform.workspace}"
-  role = aws_iam_role.teamserver_instance_role.name
+  name = "teamserver${count.index}_instance_profile_${terraform.workspace}"
+  role = aws_iam_role.teamserver_instance_role[count.index].name
 }
 
 # The instance role
```
Comment (mcdonnnj) [inline teamserver_iam.tf]:
```suggestion
# The instance roles
```

File: teamserver_iam.tf [hunk 13]
```diff
@@ -3,50 +3,62 @@
...
   provider = aws.provisionassessment
 
   name   = "assume_delegated_role_policy"
-  role   = aws_iam_role.teamserver_instance_role.id
-  policy = data.aws_iam_policy_document.teamserver_assume_delegated_role_policy_doc.json
+  role   = aws_iam_role.teamserver_instance_role[count.index].id
+  policy = data.aws_iam_policy_document.teamserver_assume_delegated_role_policy_doc[count.index].json
 }
 
 # Attach the CloudWatch Agent policy to this role as well
```
Comment (mcdonnnj) [inline teamserver_iam.tf]:
```suggestion
# Attach the CloudWatch Agent policy to these roles as well
```

File: teamserver_iam.tf [hunk 14]
```diff
@@ -3,50 +3,62 @@
...
+  count = lookup(var.operations_instance_counts, "teamserver", 0)
+
   provider = aws.provisionassessment
 
-  role       = aws_iam_role.teamserver_instance_role.id
+  role       = aws_iam_role.teamserver_instance_role[count.index].id
   policy_arn = "arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy"
 }
 
 # Attach the SSM Agent policy to this role as well
```
Comment (mcdonnnj) [inline teamserver_iam.tf]:
```suggestion
# Attach the SSM Agent policy to these roles as well
```

File: email_sending_domain_certreadrole.tf [hunk 15]
```diff
@@ -10,5 +12,5 @@ module "email_sending_domain_certreadrole" {
   cert_bucket_name = var.cert_bucket_name
   # Certbot stores wildcard certs in a directory with the name of the
   # domain, instead of pre-pending an asterisk.
-  hostname = var.email_sending_domain
+  hostname = each.key
```
Comment (mcdonnnj) [inline email_sending_domain_certreadrole.tf]: We don't have it codified, but I think that `each.value` better represents set values (and it's the same value per [the documentation](https://www.terraform.io/docs/language/meta-arguments/for_each.html#the-each-object)).

```suggestion
  hostname = each.value
```

File: gophish_iam.tf [hunk 16]
```diff
@@ -55,16 +67,18 @@ resource "aws_iam_role_policy_attachment" "efs_mount_policy_attachment_gophish"
 ################################
 
 # Allow the Gophish instance to assume the necessary role to read
```
Comment (mcdonnnj) [inline gophish_iam.tf]:
```suggestion
# Allow each Gophish instance to assume the necessary role to read
```

File: teamserver_iam.tf [hunk 17]
```diff
@@ -55,16 +67,18 @@ resource "aws_iam_role_policy_attachment" "efs_mount_policy_attachment_teamserve
 ################################
 
 # Allow the teamserver instance to assume the necessary role to read
```
Comment (mcdonnnj) [inline teamserver_iam.tf]:
```suggestion
# Allow each Teamserver instance to assume the necessary role to read
```

**Review by mcdonnnj — APPROVED** [review]
Strong work 💪💪💪. I would like us to come to a consensus on how we name non-singular resources like roles and policies, and if it would result in a change I'd like an issue created to track the work (so it's not holding up this PR). I did think it was straightforward enough to update comments to represent the fact that things are now plural, so please see my suggestions for your consideration.

File: email_sending_domain_certreadrole.tf (hunk 15 above)
Comment (dav3r) [reply email_sending_domain_certreadrole.tf]: I don't care one way or the other on this, so I'm fine with your suggestion here.

**Comment by dav3r** [comment]
> I would like us to come to a consensus on how we name non-singular resources like roles and policies, and if it would result in a change I'd like an issue created to track the work (so it's not holding up this PR).

Please take a look at https://github.com/cisagov/development-guide/issues/63 - feel free to add/update/remove content in that issue as you see fit.

**Pull request merged**